Add basic authentication

ianyong opened the pull request:
Authentication via JWT.

Note that we retrieve the user ID when we perform authentication and add it to the context. This is so as to allow the RPC method call to make use of the user ID without requiring the frontend to send the user ID in the request. An example of this can be seen in `GetPendingCollectionForUser` in `internal/adapters/primary/handlers/internal/gemservice.go` where the user ID that is used depends on the user that is logged in.

## Testing

Drop the database and set it up again by running `make dropdb` and `make setupdb`. Once that is done, navigate to the home page, where the API call will fail due to not being logged in. Then, navigate to `/login` and login as `tommy` with password `password`. You should be redirected to the home page where 6 gems should load. Repeat the above as `bobby` and you should only see 1 gem loading this time.

## To-dos

* Persisting access token in ~~local storage~~ a HTTP-only cookie (thanks @dorcastan for pointing out that local storage is susceptible to cross-site scripting attacks)
* Registration
* Making pages for the Vue application redirect to the login page if not logged in
  * Might need to devise a better abstraction for this as opposed to the current handling of auth state in `AuthInterceptor`

ianyong committed:
Add gRPC login method

ianyong committed:
Extract server initialisation into its own package

ianyong committed:
Add gRPC authentication interceptors

ianyong committed:
Add UserClaims

ianyong committed:
Add reading of SECRET_KEY from .env

ianyong committed:
Add User domain model

ianyong committed:
Add JWTManager

ianyong committed:
Remove go-grpc-middleware library

ianyong committed:
Add AuthInterceptor

ianyong committed:
Add secret key minimum length check

ianyong committed:
Add JWT authorization logic

okg00dbye reviewed (commented):
possible to add RegisterRequest/Response as well?

ianyong commented:
> possible to add RegisterRequest/Response as well?

Yup, once I figure out how to get this working HAHA

ianyong committed:
Merge branch 'master' into ian/add-basic-authentication

ianyong committed:
Add ServerStreamWithContext

ianyong committed:
Add logging interceptor

ianyong committed:
Extract out API

ianyong committed:
Add client auth interceptor

ianyong committed:
Allow 'authorization' header through reverse proxy

ianyong committed:
Set access token on frontend after logging in

ianyong committed:
Update users table & seeds

Using the past migrations so that we don't need to deal with
handling the NOT NULL constraints for old values.

ianyong committed:
Add user repository and implementation

ianyong committed:
Add login functionality

ianyong committed:
Add AuthProvider

ianyong committed:
Remove hiding of actual error

ianyong committed:
Merge branch 'master' into ian/add-basic-authentication

ianyong committed:
Add login logic to login page

ianyong committed:
Make /authapi.AuthService/Login RPC method public

ianyong committed:
Use HMAC for JWT signing

ianyong committed:
Fix authorization header format

ianyong committed:
Update GetPendingCollectionByUser RPC method to use context instead

ianyong committed:
Update seeds to give Bobby a gem

sebastiantoh commented on backend/internal/server/serverstream/serverstream.go:
```diff
@@ -0,0 +1,29 @@
+)
+
+// ServerStreamWithContext is a grpc.ServerStream that allows its context to be modified.
+// It does so by overriding the Context() function of grpc.ServerStream.
```
nit, but should this be overwrite?

```suggestion
// It does so by overwriting the Context() function of grpc.ServerStream.
```

ianyong replied:
Ah yea, I think most people use these terms loosely, but I think you're right given that `override` seems to be an OOP concept.

sebastiantoh commented on frontend/src/pages/LoginRegister/LoginPage.vue:
```diff
@@ -24,6 +24,34 @@
+    handleLogin() {
+      this.login(this.username, this.password)
+        .then(() => this.$router.push(HOME_ROUTE))
+        .catch((err) => console.log(err, "Failed to log in"));
```
maybe add a TODO to display a toast / popup to inform users that log in failed?

ianyong replied:
Oops, I didn't see the `TODO` part when I read this comment before the AGM, otherwise I would've done so already HAHA

sebastiantoh commented on frontend/src/providers/GemProvider.vue:
```diff
@@ -25,7 +22,7 @@ const protoGemToGemMapper = (protoGem: ProtoGem): Gem => {
       lng: protoGem.getLongitude(),
     },
-    createdAt: dayjs(protoGem.getCreatedAt().toDate()),
+    createdAt: dayjs(protoGem.getCreatedAt()?.toDate()),
```
~~hmm, is this necessary?~~ nvm, just saw the eslint error

ianyong replied:
I think this type mismatch was already the case, just that it didn't surface until I added `@types/google-protobuf` so that `frontend/src/api/interceptors/auth.ts` wouldn't be untyped.

With that said, should probably check if there's a not nullable version of the ProtoBuf type if we have the time. Having the timestamps be nullable removes some type safety.

sebastiantoh replied:
i don't think there is. Protobuf really emphasises backward compatibility, so all fields have to be optional iirc

ianyong committed:
Update phrasing of ServerStreamWithContext documentation

Co-authored-by: sebastiantoh <sebas.tsj.98@gmail.com>

ianyong committed:
Merge branch 'master' into ian/add-basic-authentication

ianyong committed:
Add TODO to implement login failure feedback

sebastiantoh reviewed (approved):
lgtm!
